backend/src/main/java/com/notbadcode/todo/task/TaskServiceImpl.java

```diff
@@ -0,0 +1,133 @@
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
+
+@Service
+@Transactional(readOnly = true)
+public class TaskServiceImpl implements TaskService {
+
+  private final Sort TASK_SORT = Sort.by("created").descending();
+
+  private final JsonMapper<Task> jsonMapper;
+
+  private final TaskRepository repository;
+
+  @Autowired
+  public TaskServiceImpl(TaskRepository repository,
+                         JsonMapper<Task> jsonMapper) {
+    this.repository = repository;
+    this.jsonMapper = jsonMapper;
+  }
+
+  @Override
+  @Transactional
+  public TaskDto createTaskFromString(String json) {
+    Task task = jsonMapper.fromJson(json, Task.class);
+    if (task.getTitle().isBlank()) {
+      throw new BadRequestException("title is required");
+    }
+    repository.save(task);
+    return TaskMapper.toTaskDto(task);
+  }
+
+  @Override
+  public TaskDto findTaskById(Long id) {
+    return TaskMapper.toTaskDto(getTaskById(id));
+  }
+
+  @Override
+  public List<TaskDto> findAllByFilter(String filter) {
+    TaskFilter taskFilter = TaskFilter.fromString(filter);
+    List<Task> tasks = switch (taskFilter) {
+      case ACTIVE -> repository.findAll(isCompleted(false), TASK_SORT);
+      case COMPLETED -> repository.findAll(isCompleted(true), TASK_SORT);
+      default -> repository.findAll(TASK_SORT);
+    };
+    return tasks.stream().map(TaskMapper::toTaskDto).toList();
+  }
+
+  @Override
+  @Transactional
+  public List<TaskDto> toggleAll() {
+    boolean isAllTasksCompleted = repository.existsAllByCompletedIsFalse();
+    return repository.findAll(TASK_SORT).stream()
+        .map(task -> {
+              task.setCompleted(isAllTasksCompleted);
+              return TaskMapper.toTaskDto(task);
+            })
+        .toList();
+  }
+
+  @Override
+  @Transactional
+  public TaskDto toggleTaskById(Long id) {
+    Task task = getTaskById(id);
+    task.setCompleted(!task.getCompleted());
+    return TaskMapper.toTaskDto(task);
+  }
+
+  @Override
+  @Transactional
+  public TaskDto updateTaskFromString(String json) {
+    Task newTask = jsonMapper.fromJson(json, Task.class);
+    Task task = getTaskById(newTask.getId());
+    if (newTask.getTitle() != null && !newTask.getTitle().isBlank()) {
+      task.setTitle(newTask.getTitle());
+    }
+    if (newTask.getCompleted() != null) {
+      task.setCompleted(newTask.getCompleted());
+    }
+    return TaskMapper.toTaskDto(task);
+  }
+
+  @Override
+  public TasksInfoDto getTasksInfo(String filter) {
+    TaskFilter taskFilter = TaskFilter.fromString(filter);
+    return switch (taskFilter) {
+      case ACTIVE -> buildCommonTasksInfo(repository.count(isCompleted(false)));
+      case COMPLETED -> buildCommonTasksInfo(repository.count(isCompleted(true)));
+      default -> buildCommonTasksInfo(repository.count());
+    };
```

Comment:
Можно кстати это прикольно на Optional переписать, только enum тогда надо будет немного доработать
```suggestion
    return buildCommonTasksInfo(
        taskFilter.getCompletedOptional()
            .map(completed -> repository.count(isCompleted(completed)))
            .orElseGet(repository::count));
```